Add support for localized TimeZone names

alonsodomin opened the pull request:
Fixes #44. Just a _work in progress_ submission for enabling code review and discussion. The changes are focused on just the minimal effort needed to have #44 fixed and could benefit from some optimisations (i.e. some caching around the time zone localized names).

The small change in the `build.sbt` file is to ignore `.DS_Store` files that Macs usually create in folders when you navigate them, that was making compilation break when generating the `java.time` classes since the `replaceAll` functions where invoked in a binary file.

cquiroz commented on build.sbt:
```diff
@@ -105,7 +105,7 @@ def copyAndReplace(srcDirs: Seq[File], destinationDir: File): Seq[File] = {
   val generatedFiles: List[java.io.File] = onlyScalaDirs.foldLeft(Set.empty[File]) { (files, sourceDir) =>
     files ++ copyDirectory(sourceDir, destinationDir, overwrite = true)
-  }.filterNot(_.isDirectory).filterNot(_.getParentFile.getName == "util").toList
+  }.filterNot(_.isDirectory).filter(_.getName.endsWith(".scala")).filterNot(_.getParentFile.getName == "util").toList
```
Is this change to avoid translating the `.DS_store` and other binary files?

alonsodomin replied:
yes, this is exactly for that. It could also be achieved in the `copyDirectory` method by passing a filter that only takes `.scala` files. Something in the likes of `PathFinder(source).**(scalaFileFilter)` but  that may make the build skip copying resources that could be needed afterwards, so I thought that skipping non-scala files before doing the `replaceAll` could suffice.

No problem of rolling this back if you think that the build should not take care of it...

cquiroz replied:
This is fine just wanted to make sure, this is probably faster anyway

cquiroz commented on js/src/main/scala/java/util/TimeZone.scala:
```diff
@@ -52,11 +55,16 @@ object TimeZone {
-  def getAvailableIDs(offsetMillis: Int): Array[String] = ???
+  def getAvailableIDs: Array[String] = ZoneRulesProvider.getAvailableZoneIds.asScala.toArray
+  def getAvailableIDs(offsetMillis: Int): Array[String] =
+    getAvailableIDs.view.map(getTimeZone).filter(_.getRawOffset == offsetMillis).map(_.getID).toArray
```
why the `view` and `toArray` Couldn't you transform the array directly?

alonsodomin replied:
The purpose of the `view` here is to fuse the `map(...).filter(...).map(...)` in one single traverse instead of three... but yeah, I could operate in the array straight.

cquiroz replied:
I'm not sure we use `view` in other places and a side effect is that we'll increase the `binary` size by importing the view classes.

At some moment after 2.0 I want to find out these cases to reduce the size of the export

alonsodomin replied:
right, that makes sense. At this point this is just a convenience as the same can be achieved with a single operation.

cquiroz commented on js/src/main/scala/java/util/TimeZone.scala:
```diff
@@ -84,7 +92,22 @@ abstract class TimeZone extends Serializable with Cloneable {
-    ???
+
+    val id = getID
+    val zoneName = DateFormatSymbols.getInstance(locale).getZoneStrings.find(_(0) == id).flatMap { strs =>
```
This is nice. I don't think we have this data en `scala-java-locales` but if we had it it would be good to get it

cquiroz replied:
The call `_(0)` could raise an exception. Could you handle it, e.g. filtering out empty arrays?

alonsodomin replied:
sure, will address that.

cquiroz commented:
Thanks for this. It is starting to look good.

I'm still trying to reproduce the issue with the standard test suite but extra tests are welcome too

alonsodomin commented:
Thanks, will try to add some tests around this to have something more complete.

alonsodomin commented:
Addressed previous comments and added a few tests. As you mentioned before, there is no locale information for time zones so the test in which I'm asserting over some of the display names fails (`getDisplayName` is always returning `null`). Probably worth adding that data to `scala-java-locale` (if possible) and wait before that is done before merging this.

cquiroz commented:
I probably won't have time to add the names for the timezones to `scala-java-locales` but could you add a ticket there to remind us?

cquiroz commented on js/src/main/scala/java/util/TimeZone.scala:
```diff
@@ -82,9 +90,34 @@ abstract class TimeZone extends Serializable with Cloneable {
+
+    val id = getID
+    def currentIdStrings(strs: Array[String]): Boolean =
+      atIndex(strs, 0).contains(id)
```
This is failing on travis. It seems `contains` is not supported on 2.10. Could you change this to e.g. `exists`?

cquiroz commented:
It looks very nice indeed. Have you tested it in your application?

alonsodomin commented:
Sorry for that 2.10 compile error, forgot to cross-compile it. The `NotImplemented` error that I used to get in my application is now gone and instead I get a `null` string for the zone part when formatting the timezones.

This is using a custom formatting pattern (`E, MMM d, HH:mm:ss z`), if I use the default `ISO_ZONED_DATE_TIME` I get the zone ID in it's place (should have included this bit of information in the original ticket to make it easier for you to reproduce).

So, in essence this can be considered as fixed. I believe the build will still not pass due to test failures so if you prefer to delegate the missing bits towards the locales project I can remove/comment out the failing tests and consider this bit of work as complete.

cquiroz commented:
Thanks a lot for your contribution. Could you rebase on top of the latest master?

The failing tests seems to be due to the missing locales data. Could you mark it as ignored or pending? It is certainly worth having it here

alonsodomin commented:
Rebased on top of master and ignoring the display name test for now. This should be ready for merging after a successful build.

cquiroz commented:
Thanks a lot this seems ready.
